src/control/system/raft/raft_recovery_test.go

```diff
@@ -0,0 +1,471 @@
+//
+// (C) Copyright 2022 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package raft
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/google/uuid"
+	"github.com/hashicorp/raft"
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/common"
+	"github.com/daos-stack/daos/src/control/common/test"
+	"github.com/daos-stack/daos/src/control/logging"
+	. "github.com/daos-stack/daos/src/control/system"
+)
+
+var regenRaftFixtures = flag.Bool("regen-raft-fixtures", false, "regenerate raft test files")
+
+const (
+	raftFixDir = "testdata/raft_recovery"
+)
+
+func waitForSnapshots(ctx context.Context, t *testing.T, log logging.Logger, cfg *DatabaseConfig, num int) {
+	t.Helper()
+	for {
+		select {
+		case <-ctx.Done():
+			t.Fatal(ctx.Err())
+		default:
+			snaps, err := GetSnapshotInfo(log, cfg)
+			if err != nil && err != ErrNoRaftSnapshots {
+				t.Fatal(err)
+			}
+			if len(snaps) >= num {
+				return
+			}
+			time.Sleep(1 * time.Second)
+		}
+	}
+}
+
+func testDbCfg() *DatabaseConfig {
+	return &DatabaseConfig{
+		Replicas:              []*net.TCPAddr{common.LocalhostCtrlAddr()},
+		SystemName:            "daos-test",
+		RaftDir:               raftFixDir,
+		RaftSnapshotInterval:  1 * time.Second,
+		RaftSnapshotThreshold: 8,
+	}
+}
+
+func Test_Raft_RegenerateFixtures(t *testing.T) {
+	if !*regenRaftFixtures {
+		return
+	}
+
+	if err := os.RemoveAll(raftFixDir); err != nil {
+		t.Fatal(err)
+	}
```

Comment:
This test was really confusing to me at first. Took some study to figure out this must be manually turned on to regenerate the testdata directory. I'm still not sure why we'd want to do that, though.

Reply:
I didn't want to run the risk of things being unstable by regenerating the test fixtures all of the time. With the current fixtures, we have two snapshots at specific points in time with clear expectations about what the system should look like after restoring from those snapshots.

I considered using `t.Skip()` with a message, but that seems to cause heartburn during release testing, so I just opted to have the test return early when we're not regenerating the fixtures manually.

Reply:
I was wondering why we'd want to regen the fixtures at all, I mean, short of making an actual change in them, which I expect would be rare.

Reply:
Well, as an example, I needed to regenerate them after merging this back into the catastrophic recovery branch because the size of the MS DB is slightly larger on that branch. As a result, the snapshot recovery unit tests were failing when they compared the reconstituted DB contents to the snapshot contents.

Reply:
Ah, fair point.